### ISSUE TITLE
graphql-ruby-client doesn't play well with custom authorization strategies

### ISSUE THREAD

**Issue by kevinjalbert** [rmosolgo/graphql-ruby#959]
I ran into this situation when setting up [graphql-ruby-client](https://github.com/rmosolgo/graphql-ruby-client) with graphql-pro.

I have a [Custom Authorization Strategy](http://graphql-ruby.org/pro/authorization.html#custom-authorization-strategy) for my schema, which looks for a specific auth bearer token.

When running the `graphql-ruby-client` the sync will fail on the server with the following backtrace:

```
NoMethodError (undefined method `allowed?' for nil:NilClass):

graphql-pro (1.5.3) lib/graphql/pro/access/schema_filter.rb:32:in `call'
graphql (1.6.8) lib/graphql/filter.rb:13:in `call'
graphql (1.6.8) lib/graphql/filter.rb:12:in `call'
graphql (1.6.8) lib/graphql/schema/warden.rb:43:in `block in initialize'
graphql (1.6.8) lib/graphql/schema/warden.rb:138:in `block in read_through'
graphql (1.6.8) lib/graphql/schema/warden.rb:134:in `visible?'
graphql (1.6.8) lib/graphql/schema/warden.rb:120:in `root_type_for_operation'
graphql (1.6.8) lib/graphql/static_validation/rules/mutation_root_exists.rb:8:in `validate'
graphql (1.6.8) lib/graphql/static_validation/validator.rb:34:in `block in validate'
graphql (1.6.8) lib/graphql/static_validation/validator.rb:33:in `each'
graphql (1.6.8) lib/graphql/static_validation/validator.rb:33:in `validate'
graphql-pro (1.5.3) lib/graphql/pro/operation_store/validate.rb:20:in `validate'
graphql-pro (1.5.3) lib/graphql/pro/operation_store/add_operation.rb:14:in `call'
graphql-pro (1.5.3) lib/graphql/pro/operation_store.rb:101:in `add'
graphql-pro (1.5.3) lib/graphql/pro/operation_store/endpoint.rb:40:in `block (2 levels) in call'
graphql-pro (1.5.3) lib/graphql/pro/operation_store/endpoint.rb:37:in `each'
graphql-pro (1.5.3) lib/graphql/pro/operation_store/endpoint.rb:37:in `block in call'
activerecord (5.1.2) lib/active_record/connection_adapters/abstract/database_statements.rb:225:in `block in transaction'
activerecord (5.1.2) lib/active_record/connection_adapters/abstract/transaction.rb:194:in `block in within_new_transaction'
/Users/jalbert/.rvm/rubies/ruby-2.4.1/lib/ruby/2.4.0/monitor.rb:214:in `mon_synchronize'
activerecord (5.1.2) lib/active_record/connection_adapters/abstract/transaction.rb:191:in `within_new_transaction'
activerecord (5.1.2) lib/active_record/connection_adapters/abstract/database_statements.rb:225:in `transaction'
activerecord (5.1.2) lib/active_record/transactions.rb:210:in `transaction'
graphql-pro (1.5.3) lib/graphql/pro/operation_store/endpoint.rb:35:in `call'
actionpack (5.1.2) lib/action_dispatch/routing/mapper.rb:17:in `block in <class:Constraints>'
actionpack (5.1.2) lib/action_dispatch/routing/mapper.rb:46:in `serve'
actionpack (5.1.2) lib/action_dispatch/journey/router.rb:46:in `block in serve'
actionpack (5.1.2) lib/action_dispatch/journey/router.rb:33:in `each'
actionpack (5.1.2) lib/action_dispatch/journey/router.rb:33:in `serve'
actionpack (5.1.2) lib/action_dispatch/routing/route_set.rb:832:in `call'
rack (2.0.3) lib/rack/etag.rb:25:in `call'
rack (2.0.3) lib/rack/conditional_get.rb:38:in `call'
rack (2.0.3) lib/rack/head.rb:12:in `call'
activerecord (5.1.2) lib/active_record/migration.rb:556:in `call'
actionpack (5.1.2) lib/action_dispatch/middleware/callbacks.rb:26:in `block in call'
activesupport (5.1.2) lib/active_support/callbacks.rb:97:in `run_callbacks'
actionpack (5.1.2) lib/action_dispatch/middleware/callbacks.rb:24:in `call'
actionpack (5.1.2) lib/action_dispatch/middleware/executor.rb:12:in `call'
bugsnag (5.3.2) lib/bugsnag/rack.rb:34:in `call'
actionpack (5.1.2) lib/action_dispatch/middleware/debug_exceptions.rb:59:in `call'
actionpack (5.1.2) lib/action_dispatch/middleware/show_exceptions.rb:31:in `call'
railties (5.1.2) lib/rails/rack/logger.rb:36:in `call_app'
railties (5.1.2) lib/rails/rack/logger.rb:24:in `block in call'
activesupport (5.1.2) lib/active_support/tagged_logging.rb:69:in `block in tagged'
activesupport (5.1.2) lib/active_support/tagged_logging.rb:26:in `tagged'
activesupport (5.1.2) lib/active_support/tagged_logging.rb:69:in `tagged'
railties (5.1.2) lib/rails/rack/logger.rb:24:in `call'
actionpack (5.1.2) lib/action_dispatch/middleware/remote_ip.rb:79:in `call'
request_store (1.3.2) lib/request_store/middleware.rb:9:in `call'
actionpack (5.1.2) lib/action_dispatch/middleware/request_id.rb:25:in `call'
rack (2.0.3) lib/rack/runtime.rb:22:in `call'
activesupport (5.1.2) lib/active_support/cache/strategy/local_cache_middleware.rb:27:in `call'
actionpack (5.1.2) lib/action_dispatch/middleware/executor.rb:12:in `call'
actionpack (5.1.2) lib/action_dispatch/middleware/static.rb:125:in `call'
rack (2.0.3) lib/rack/sendfile.rb:111:in `call'
rack-cors (0.4.1) lib/rack/cors.rb:81:in `call'
railties (5.1.2) lib/rails/engine.rb:522:in `call'
rack (2.0.3) lib/rack/handler/webrick.rb:86:in `service'
/Users/jalbert/.rvm/rubies/ruby-2.4.1/lib/ruby/2.4.0/webrick/httpserver.rb:140:in `service'
/Users/jalbert/.rvm/rubies/ruby-2.4.1/lib/ruby/2.4.0/webrick/httpserver.rb:96:in `run'
/Users/jalbert/.rvm/rubies/ruby-2.4.1/lib/ruby/2.4.0/webrick/server.rb:290:in `block in start_thread'
```

If I remove my custom authorization strategy everything works as expected.

-----

Some extra information from direct messages with @rmosolgo:
> I bet `sync` doesn’t play well with `view:` because `view:` requires `current_user` during validation

**Comment by rmosolgo** [comment]
Thanks for opening this issue! 

The trick is: 

- When persisting queries, we validate them first 
- If you're using `view:`, then a `current_user:` is required for validation. 
- But, when persisting queries, there _is no_ `current_user` (the app is not running; it's a build task)

So, I think the solution is to pass the current `Client` as `current_user:`. For example, if you're persisting queries for your `iOS` app, you would get 

```ruby
puts context[:current_user]
# #<GraphQL::Pro...::GraphQLClient name="iOS-v1.1">
context[:current_user].name 
# "iOS-v1.1"
```

Then, update the validation logic to handle the incoming client (ie, should _all_ `iOS-v1.1` users be allowed to see this field/type?), for example: 

```ruby 
def allowed?(gate, obj) 
  case @current_user 
  when User 
    # normal user-based auth logic 
  when GraphQL::Pro::OperationStore::Client 
     # authorize the client application 
  end 
end 
```

How does that sound? 

(Another way to implement the same logic is to switch during `#initialize`,  for example 

```ruby 
def initialize(context)
  current_user_or_client = context[:current_user]
  if current_user_or_client.is_a?(User)
    @current_user = current_user_or_client 
  else 
    @current_client_app = current_user_or_client 
  end 
end 

def allowed?(gate, obj)
  if @current_user 
    # ...
  elsif @current_client_app 
    # ...
  else 
    raise "Can't authorize without app or user"
  end 
end 
```
)

**Comment by kevinjalbert** [comment]
@rmosolgo that makes sense.

Unfortunately, I'm running into an issue trying this out. I have `debugger`s put as the first statement in all my methods within my Custom Authorization Strategy class, and nothing is triggering it when I use `graphql-ruby-client sync`. It seems like the graphql-pro side of things is happening first before I even get to my authorization class?

I'm using graphql version 1.6.8 and graphql-pro version 1.5.3 if that helps.

**Comment by rmosolgo** [comment]
Oh, that's a good point -- it's erroring _before_ the `#allowed?` call, isn't it? 

It's a problem _before_ the problem :P Since `sync` uses `#validate` but doesn't actually run the query, the required `#before_query` hook is not being run, so the context isn't ready. 

I'll have to add a hook for that! Let me take a look...

**Comment by rmosolgo** [comment]
I worked this up last night but took a different track: 

The problem with the approach that I mentioned above is that it requires you to modify _all_ your existing auth logic to support `OperationStore`-related logic. That's a really annoying problem because it's tedious and it increases the chance that you'd _accidentally_ mess up your existing auth logic and introduce a security hole. 

So instead, I implemented a _separate_ auth strategy for persisted operations. You can implement a separate strategy which will be used for incoming persisted operations. That way, similar logic stays grouped together and you don't have to do "shotgun surgery" on your existing code. 

It's released in graphql-pro 1.5.6, and you can see the docs here: http://graphql-ruby.org/operation_store/access_control.html#authorization

That'll take care of the runtime error described in this issue, so when you get a chance to try it out, please open an issue if it's not quite right or if I can help out at all!
